# Wrangle Pisa 2012
#### By Gabriela Sikora

Throughout this notebook, I will be assessing and cleaning data from the PISA 2012 dataset. The purpose of this wrangling is to have a usable dataframe for the complementary exploration notebook, wherein I will be exploring the cleaned dataframe and drawing conclusions from the plots.


### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#references">References</a></li>    
</ul>

<a id='intro'></a>
## Introduction



To begin, we first need to import all the relevant libraries:
#### Import

In [1]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [2]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

/Users/gabriela/Desktop/Code/myenv/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [3]:
# Quick overview of the entire dataframe
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [4]:
# Check how many entries and types of data
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [5]:
# Look through the datatypes
pisa.dtypes[400:]

ST23Q08        object
ST24Q01        object
ST24Q02        object
ST24Q03        object
CLCUSE1        object
CLCUSE301       int64
CLCUSE302       int64
DEFFORT         int64
QUESTID        object
BOOKID         object
EASY           object
AGE           float64
GRADE         float64
PROGN          object
ANXMAT        float64
ATSCHL        float64
ATTLNACT      float64
BELONG        float64
BFMJ2         float64
BMMJ1         float64
CLSMAN        float64
COBN_F         object
COBN_M         object
COBN_S         object
COGACT        float64
CULTDIST      float64
CULTPOS       float64
DISCLIMA      float64
ENTUSE        float64
ESCS          float64
               ...   
W_FSTR55      float64
W_FSTR56      float64
W_FSTR57      float64
W_FSTR58      float64
W_FSTR59      float64
W_FSTR60      float64
W_FSTR61      float64
W_FSTR62      float64
W_FSTR63      float64
W_FSTR64      float64
W_FSTR65      float64
W_FSTR66      float64
W_FSTR67      float64
W_FSTR68      float64
W_FSTR69  

In [6]:
# Check dictionary for names of variables
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [7]:
# Check for null values in relevant columns
columns = ['CNT', 'STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

CNT 0
STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [8]:
# See countries included in dataframe
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [9]:
# Check if there are any duplicates
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
3. Make the variable names readable since they are codes that can only be read with the dictionary

4. There are multiple columns for single variables related to father's education
5. There are multiple columns for single variables related to mother's education

6. Missing values in relevant columns

7. Create a total for Out-of-School Study Time and Learning time (minutes per week)
8. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)


<a id='cleaning1'></a>
### First Cleaning

In [10]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### 1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed


There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [11]:
# Create new dataframe with only the selected columns
pisa_clean = pisa_clean[['CNT', 'STIDSTD', 'ST04Q01', 'ST13Q01', 'ST14Q01', 'ST14Q02', 'ST14Q03', 'ST14Q04', 
                         'ST17Q01', 'ST18Q01', 'ST18Q02', 'ST18Q03', 'ST18Q04', 
                         'ST57Q01', 'ST57Q02', 'ST57Q03', 'ST57Q04', 'ST57Q05', 'MMINS', 
                         'LMINS', 'SMINS', 'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 
                         'PV5MATH', 'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [12]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['CNT',
 'STIDSTD',
 'ST04Q01',
 'ST13Q01',
 'ST14Q01',
 'ST14Q02',
 'ST14Q03',
 'ST14Q04',
 'ST17Q01',
 'ST18Q01',
 'ST18Q02',
 'ST18Q03',
 'ST18Q04',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### 2. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [13]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score']) / 3

In [14]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [15]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())
print('-'*22)

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
----------------------
count    485490.000000
mean        472.488764
std          96.036271
min          77.114593
25%         403.992595
50%         472.046460
75%         541.455700
max         826.592027
Name: Average Total Score, dtype: float64


In [16]:
# Test to see if the data from the new variables display correctly
pisa_clean.sample(10)

CNT  STIDSTD ST04Q01            ST13Q01 ST14Q01 ST14Q02  \
171063      Estonia      168  Female  <ISCED level 3A>       No      No   
468398       Turkey     3009    Male   <ISCED level 1>       No      No   
469476       Turkey     4087  Female   <ISCED level 1>       No      No   
390216        Qatar      966    Male  <ISCED level 3A>       No     Yes   
269205        Italy    28616    Male                NaN     NaN     NaN   
461983      Tunisia     1001    Male                NaN      No      No   
92847        Canada    18378    Male  <ISCED level 3A>      Yes     NaN   
239638       Israel     4104  Female  <ISCED level 3A>      NaN     NaN   
98131   Switzerland     2118    Male   <ISCED level 1>      NaN     NaN   
385510     Portugal     1982    Male  <ISCED level 3A>       No      No   

       ST14Q03 ST14Q04            ST17Q01 ST18Q01  ... ST57Q03 ST57Q04  \
171063      No     Yes  <ISCED level 3A>       No  ...     0.0     0.0   
468398      No     NaN  <ISCED level 3A>       No  ...     NaN     NaN   
469476      No     NaN   <ISCED level 1>       No  ...     NaN     NaN   
390216      No      No  <ISCED level 3A>       No  ...     1.0     3.0   
269205     NaN     NaN                NaN     NaN  ...     NaN     NaN   
461983      No     Yes                NaN      No  ...     0.0     0.0   
92847      NaN     NaN  <ISCED level 3A>      Yes  ...     0.0     0.0   
239638     NaN     NaN  <ISCED level 3A>      NaN  ...     NaN     0.0   
98131      NaN     NaN   <ISCED level 1>      NaN  ...     0.0     0.0   
385510      No      No   <ISCED level 2>       No  ...     NaN     NaN   

       ST57Q05  MMINS  LMINS  SMINS  Average Math Score  \
171063     2.0  270.0  180.0  360.0           470.56398   
468398     NaN    NaN    NaN    NaN           431.14972   
469476     NaN    NaN    NaN    NaN           452.95996   
390216     2.0  275.0    NaN    NaN           358.86430   
269205     NaN    NaN    NaN    NaN           443.30116   
461983     3.0  240.0  300.0   60.0           360.26638   
92847      0.0  375.0  375.0  375.0           580.78362   
239638     4.0  225.0  135.0   90.0           519.09176   
98131      1.0  225.0  225.0  270.0           450.46734   
385510     NaN    NaN    NaN    NaN           564.50384   

        Average Reading Score  Average Science Score  Average Total Score  
171063              480.95872              484.83778           478.786827  
468398              463.66640              454.34538           449.720500  
469476              511.46024              462.64456           475.688253  
390216              373.68816              336.29242           356.281627  
269205              345.45968              484.93100           424.563947  
461983              365.82910              402.40582           376.167100  
92847               459.01512              525.40094           521.733227  
239638              449.42460              486.79594           485.104100  
98131               434.63596              406.78850           430.630600  
385510              506.57044              522.69674           531.257007  

[10 rows x 25 columns]

### 3. Make the variable names readable since they are codes that can only be read with the dictionary

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [17]:
# Rename columns
pisa_clean.rename({'CNT': 'Country',
                   'STIDSTD': 'Student ID',
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother - Highest Schooling', 
                   'ST14Q01': 'Mother Qualifications - ISCED level 6',
                   'ST14Q02': 'Mother Qualifications - ISCED level 5A',
                   'ST14Q03': 'Mother Qualifications - ISCED level 5B',
                   'ST14Q04': 'Mother Qualifications - ISCED level 4',
                   'ST17Q01': 'Father - Highest Schooling',
                   'ST18Q01': 'Father Qualifications - ISCED level 6',
                   'ST18Q02': 'Father Qualifications - ISCED level 5A',
                   'ST18Q03': 'Father Qualifications - ISCED level 5B',
                   'ST18Q04': 'Father Qualifications - ISCED level 4',
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning Time - Mathematics',
                   'LMINS': 'Learning Time - Test Language',
                   'SMINS': 'Learning Time - Science'}, axis='columns', inplace=True)

In [18]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Country',
 'Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'Father Qualifications - ISCED level 6',
 'Father Qualifications - ISCED level 5A',
 'Father Qualifications - ISCED level 5B',
 'Father Qualifications - ISCED level 4',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score']

### 4. There are multiple columns for single variables related to father's education

Values for the father's education are distributed among 5 columns. One column holds '<ISCED level 1>', '<ISCED level 2>', '<ISCED level 3A>', '<ISCED level 3B, 3C>', and 'He did not complete <ISCED level 1>'. We would rather have one level for ISCED level 3, and change 'He did not complete <ISCED level 1>' to level 0, in accordance to this classification: http://uis.unesco.org/sites/default/files/documents/international-standard-classification-of-education-isced-2011-en.pdf

Also, since the data for this variable has multiple pieces of data in one column, and then another 4 columns hold the rest of the data, we will have to expand the first column by creating dummies for each data to have a column, and then melt all the columns into one. 

In [19]:
# Begin with a copy
father_edu = pisa_clean.copy()

In [20]:
# Split column up with dummies
father_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'He did not complete <ISCED level 1>']] = pd.get_dummies(father_edu['Father - Highest Schooling'])

In [21]:
# Rename the columns for consistency and to remove subcategorization of levels
father_edu.rename({'He did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [22]:
# Change datatypes from ints into strings for dummy columns
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].astype(str)
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].astype(str)
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].astype(str)
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].astype(str)

In [23]:
# Change values from '1's and 'Yes's to the corresponding level to prepare for the melt
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
father_edu['<ISCED level 4>'] = father_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
father_edu['<ISCED level 5>'] = father_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
father_edu['<ISCED level 6>'] = father_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [24]:
# Test to see if the columns are all correct before melt
father_edu.sample(10)

Country  Student ID  Gender Mother - Highest Schooling  \
259102            Italy       18513    Male           <ISCED level 2>    
277022           Jordan        5360    Male          <ISCED level 3A>    
36744           Austria         113  Female      <ISCED level 3B, 3C>    
210860  Hong Kong-China        3960    Male      <ISCED level 3B, 3C>    
218055          Hungary        1477  Female          <ISCED level 3A>    
335531           Mexico       19829  Female          <ISCED level 3A>    
345797           Mexico       30095    Male      <ISCED level 3B, 3C>    
182902          Finland        7228    Male          <ISCED level 3A>    
363747      Netherlands        4298  Female           <ISCED level 1>    
454529         Thailand         153  Female           <ISCED level 2>    

       Mother Qualifications - ISCED level 6  \
259102                                    No   
277022                                   NaN   
36744                                     No   
210860                                    No   
218055                                    No   
335531                                    No   
345797                                   NaN   
182902                                    No   
363747                                   NaN   
454529                                    No   

       Mother Qualifications - ISCED level 5A  \
259102                                     No   
277022                                    NaN   
36744                                      No   
210860                                     No   
218055                                     No   
335531                                     No   
345797                                    NaN   
182902                                    Yes   
363747                                    NaN   
454529                                     No   

       Mother Qualifications - ISCED level 5B  \
259102                                     No   
277022                                    Yes   
36744                                     Yes   
210860                                     No   
218055                                     No   
335531                                    Yes   
345797                                    Yes   
182902                                     No   
363747                                    NaN   
454529                                    NaN   

       Mother Qualifications - ISCED level 4 Father - Highest Schooling  \
259102                                    No           <ISCED level 2>    
277022                                   NaN          <ISCED level 3A>    
36744                                     No      <ISCED level 3B, 3C>    
210860                                    No           <ISCED level 1>    
218055                                   Yes                        NaN   
335531                                   NaN           <ISCED level 2>    
345797                                   NaN           <ISCED level 2>    
182902                                    No          <ISCED level 3A>    
363747                                   NaN          <ISCED level 3A>    
454529                                    No           <ISCED level 2>    

       <ISCED level 6>  ... Learning Time - Science Average Math Score  \
259102              No  ...                   120.0          454.67364   
277022             NaN  ...                     NaN          248.72256   
36744               No  ...                     NaN          463.94298   
210860              No  ...                   480.0          663.19520   
218055             NaN  ...                     NaN          384.72504   
335531              No  ...                   500.0          472.97868   
345797             NaN  ...                     NaN          283.46314   
182902              No  ...                   135.0          565.04908   
363747              No  ...                     0.0          459.11358   
454529              No  ..

In [25]:
# Combine all columns into one variable column
father_edu = pd.melt(father_edu, 
                     id_vars=['Country','Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='Completed_edu - Father', value_name='Education - Father')


In [26]:
# Check to see if the columns from the melt were correctly created
father_edu.sample(10)

Country  Student ID  Gender  \
921879        Slovak Republic        3384  Female   
1940188               Vietnam        3188  Female   
3006335                Canada       18926  Female   
2697285                 Italy       29246  Female   
3148175               Iceland        3209    Male   
763452                 Jordan        6300    Male   
3564924                 Spain       20912  Female   
1245700                Jordan        3058    Male   
2329011              Portugal        3523  Female   
4381160  United Arab Emirates        7008    Male   

                    Mother - Highest Schooling  \
921879                   <ISCED level 3B, 3C>    
1940188                       <ISCED level 1>    
3006335                      <ISCED level 3A>    
2697285                       <ISCED level 2>    
3148175                                    NaN   
763452                        <ISCED level 1>    
3564924                       <ISCED level 2>    
1245700                      <ISCED level 3A>    
2329011  She did not complete <ISCED level 1>    
4381160  She did not complete <ISCED level 1>    

        Mother Qualifications - ISCED level 6  \
921879                                     No   
1940188                                    No   
3006335                                    No   
2697285                                    No   
3148175                                   NaN   
763452                                     No   
3564924                                    No   
1245700                                    No   
2329011                                    No   
4381160                                    No   

        Mother Qualifications - ISCED level 5A  \
921879                                      No   
1940188                                     No   
3006335                                     No   
2697285                                     No   
3148175                                    NaN   
763452                                      No   
3564924                                     No   
1245700                                    Yes   
2329011                                     No   
4381160                                     No   

        Mother Qualifications - ISCED level 5B  \
921879                                      No   
1940188                                    NaN   
3006335                                    Yes   
2697285                                     No   
3148175                                    NaN   
763452                                      No   
3564924                                     No   
1245700                                    Yes   
2329011                                     No   
4381160                                     No   

        Mother Qualifications - ISCED level 4  \
921879                                     No   
1940188                                    No   
3006335                                    No   
2697285                                   Yes   
3148175                                   NaN   
763452                                    NaN   
3564924                                    No   
1245700                                   NaN   
2329011                                    No   
4381160                                    No   

         Out-of-School Study Time - Homework  \
921879                                   NaN   
1940188                                  4.0   
3006335                                  4.0   
2697285                                  3.0   
3148175                                  NaN   
763452                                   NaN   
3564924                                  NaN   
1245700                                  2.0   
2329011                                  2.0   
4381160                                  2.0   

         Out-of-School Study Time - Guided Homework  ...  \
921879                                          NaN  ...   
1940188                                         3.0  ...

In [27]:
# Remove all cases from melt where the education of the father is either '0' or 'No'
father_edu = father_edu[(father_edu['Education - Father'] != '0')] 
father_edu = father_edu[(father_edu['Education - Father'] != 'No')]

In [28]:
# Check to see if 0's and No's have been removed
father_edu.sample(70)

Country  Student ID  Gender  \
1991071               Belgium        7725    Male   
1973610             Australia        9500    Male   
292239                  Korea        1380  Female   
474853                Uruguay        4616  Female   
1352947                Poland        3046    Male   
3691678                 Korea        2389  Female   
325421                 Mexico        9719    Male   
3400091               Albania        1662  Female   
3702436            Luxembourg        3203  Female   
3355493              Slovenia        4870    Male   
7448     United Arab Emirates        2706  Female   
2264052                Mexico        6390  Female   
397436                  Qatar        8186    Male   
543594                 Brazil        2839  Female   
202099                 Greece         324    Male   
1516156                Brazil        4421  Female   
534874                Belgium        7998  Female   
48266                 Belgium        6880  Female   
3611210               Croatia        1210  Female   
1024942              Bulgaria        3979  Female   
3403887  United Arab Emirates         715  Female   
2373899                Serbia        3618    Male   
384430               Portugal         902  Female   
2258077                Mexico         415  Female   
721351                 Israel         327    Male   
3455349                Brazil        1654  Female   
2282908                Mexico       25246    Male   
2280593                Mexico       22931    Male   
1225809                 Italy       14240  Female   
3338273             Singapore        2558    Male   
...                       ...         ...     ...   
1310693                Mexico       24011  Female   
202734                 Greece         959    Male   
2196551                 Italy       14002  Female   
2059131              Colombia        3073  Female   
1224446                 Italy       12877    Male   
563371                 Canada        3412    Male   
390835                  Qatar        1585    Male   
661849                Finland         685  Female   
3777284                  Peru        5968  Female   
433753        Slovak Republic         748  Female   
2266433                Mexico        8771    Male   
2035119                Canada       18690    Male   
97033             Switzerland        1020    Male   
2283059                Mexico       25397    Male   
946696                Tunisia         224    Male   
2949076             Australia       13986    Male   
351602             Montenegro        2094    Male   
3639251                 Italy         232    Male   
2299688              Malaysia        3476    Male   
2276146                Mexico       18484    Male   
3181189                 Italy       27660  Female   
3617527               Hungary        2519    Male   
3629598               Ireland        4158  Female   
1772694                Mexico         522  Female   
75468                  Canada         999    Male   
574827                 Canada       14868    Male   
1163717        United Kingdom        3621    Male   
1523699                Brazil       11964  Female   
328381                 Mexico       12679    Male   
209752        Hong Kong-China        2852    Male   

                    Mother - Highest Schooling  \
1991071                  <ISCED level 3B, 3C>    
1973610                      <ISCED level 3A>    
292239                       <ISCED level 3A>    
474853                       <ISCED level 3A>    
1352947                      <ISCED level 3A>    
3691678                      <ISCED level 3A>    
325421                       <ISCED level 3A>    
3400091                      <ISCED level 3A>    
3702436                      <ISCED level 3A>    
3355493                  <ISCED level 3B, 3C>    
7448                         <ISCED level 3A>    
2264052                      <ISCED level 3A>    
397436                       <ISCED level 3A>    
543594                       <ISCED

In [29]:
# Remove 'Father - Highest Schooling' since it is redundant
father_edu = father_edu[(father_edu['Completed_edu - Father'] != 'Father - Highest Schooling')] 

In [30]:
# Check to see if correct values remain
father_edu['Education - Father'].value_counts()

<ISCED level 3>    306459
<ISCED level 5>    185130
<ISCED level 2>     84329
<ISCED level 4>     77265
<ISCED level 1>     35938
<ISCED level 6>     29589
<ISCED level 0>     16535
Name: Education - Father, dtype: int64

In [31]:
# Completed_edu is no longer needed since it duplicates the information in 'Education - Father'
father_edu.drop(columns=['Completed_edu - Father'], axis=1, inplace=True)

In [32]:
# Ensure the column has been dropped
father_edu.sample(10)

Country  Student ID  Gender Mother - Highest Schooling  \
4059317          Estonia        4502  Female          <ISCED level 3A>    
3078866            Spain       20344    Male          <ISCED level 3A>    
665039           Finland        3875  Female          <ISCED level 3A>    
1406137  Slovak Republic        2152    Male      <ISCED level 3B, 3C>    
1937348          Vietnam         348    Male           <ISCED level 1>    
3551578            Spain        7566    Male      <ISCED level 3B, 3C>    
1805988       Montenegro          10  Female      <ISCED level 3B, 3C>    
4037569            Spain        8067    Male      <ISCED level 3B, 3C>    
3821281        Singapore          76  Female          <ISCED level 3A>    
659213           Estonia        2828    Male           <ISCED level 2>    

        Mother Qualifications - ISCED level 6  \
4059317                                    No   
3078866                                    No   
665039                                    Yes   
1406137                                   NaN   
1937348                                    No   
3551578                                    No   
1805988                                    No   
4037569                                    No   
3821281                                    No   
659213                                     No   

        Mother Qualifications - ISCED level 5A  \
4059317                                     No   
3078866                                     No   
665039                                     NaN   
1406137                                    NaN   
1937348                                     No   
3551578                                     No   
1805988                                    Yes   
4037569                                     No   
3821281                                     No   
659213                                      No   

        Mother Qualifications - ISCED level 5B  \
4059317                                     No   
3078866                                     No   
665039                                     NaN   
1406137                                    Yes   
1937348                                    NaN   
3551578                                     No   
1805988                                     No   
4037569                                    Yes   
3821281                                    Yes   
659213                                      No   

        Mother Qualifications - ISCED level 4  \
4059317                                   Yes   
3078866                                    No   
665039                                    NaN   
1406137                                   NaN   
1937348                                    No   
3551578                                   Yes   
1805988                                   NaN   
4037569                                   Yes   
3821281                                   Yes   
659213                                     No   

         Out-of-School Study Time - Homework  \
4059317                                  7.0   
3078866                                  3.0   
665039                                   2.0   
1406137                                  1.0   
1937348                                  NaN   
3551578                                  4.0   
1805988                                  7.0   
4037569                                  5.0   
3821281                                  NaN   
659213                                   NaN   

         Out-of-School Study Time - Guided Homework  ...  \
4059317                                         0.0  ...   
3078866                                         0.0  ...   
665039                                          0.0  ...   
1406137                                         0.0  ...   
1937348                                         NaN  ...   
3551578                                         2.0  ...   
1805988                                         0.0  ...   


In [33]:
# Check if there are any duplicated values remaining
father_edu['Student ID'].duplicated().sum()

1229261

In [34]:
# https://stackoverflow.com/questions/33042777/removing-duplicates-from-pandas-dataframe-with-condition-for-retaining-original
# Change datatypes into category, and put into order from lowest to highest
father_edu['Education - Father'] = father_edu['Education - Father'].astype('category')
father_edu['Education - Father'] = father_edu['Education - Father'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
father_edu.sort_values(['Education - Father'], inplace=True)

# Remove duplicates and keep last, which is the highest level of education for the student's father
father_edu_clean = father_edu.drop_duplicates(keep='last')

In [35]:
# Ensure there are no duplicates
father_edu_clean['Student ID'].duplicated().sum()

901800

In [36]:
# Check the size of the dataframe - 
print(pisa_clean.shape)
print(father_edu_clean.shape)

(485490, 25)
(935606, 21)


In [37]:
# Final check of clean dataframe
father_edu_clean.sample(10)

Country  Student ID  Gender  \
3708036      Latvia        3545    Male   
2282126      Mexico       24464  Female   
1759662  Luxembourg        2389    Male   
3843830      Sweden        1806  Female   
1672048     Croatia        4008    Male   
3449088    Bulgaria         675    Male   
1958734   Argentina         532    Male   
3070405       Spain       11883    Male   
2124133     Finland        6499  Female   
3513719    Colombia        1191    Male   

                    Mother - Highest Schooling  \
3708036                      <ISCED level 3A>    
2282126                      <ISCED level 3A>    
1759662                  <ISCED level 3B, 3C>    
3843830                      <ISCED level 3A>    
1672048                      <ISCED level 3A>    
3449088                      <ISCED level 3A>    
1958734  She did not complete <ISCED level 1>    
3070405                      <ISCED level 3A>    
2124133                  <ISCED level 3B, 3C>    
3513719  She did not complete <ISCED level 1>    

        Mother Qualifications - ISCED level 6  \
3708036                                    No   
2282126                                    No   
1759662                                   NaN   
3843830                                    No   
1672048                                    No   
3449088                                    No   
1958734                                    No   
3070405                                    No   
2124133                                    No   
3513719                                    No   

        Mother Qualifications - ISCED level 5A  \
3708036                                     No   
2282126                                     No   
1759662                                    NaN   
3843830                                     No   
1672048                                     No   
3449088                                     No   
1958734                                     No   
3070405                                     No   
2124133                                     No   
3513719                                     No   

        Mother Qualifications - ISCED level 5B  \
3708036                                     No   
2282126                                     No   
1759662                                    NaN   
3843830                                    Yes   
1672048                                    Yes   
3449088                                     No   
1958734                                     No   
3070405                                    Yes   
2124133                                     No   
3513719                                     No   

        Mother Qualifications - ISCED level 4  \
3708036                                   Yes   
2282126                                   NaN   
1759662                                   NaN   
3843830                                   NaN   
1672048                                   Yes   
3449088                                    No   
1958734                                   NaN   
3070405                                    No   
2124133                                   Yes   
3513719                                   NaN   

         Out-of-School Study Time - Homework  \
3708036                                  NaN   
2282126                                  NaN   
1759662                                  NaN   
3843830                                 18.0   
1672048                                  4.0   
3449088                                  3.0   
1958734                                  0.0   
3070405                                  0.0   
2124133                                  NaN   
3513719                                  2.0   

         Out-of-School Study Time - Guided Homework  ...  \
3708036                                         NaN  ...   
2282126                                         NaN  ...   
1759662                                         NaN  ...   
3843830                             

### 5. There are multiple columns for single variables related to mother's education

The exact same process for the mother's education is followed as for the father's education. 

In [38]:
# Build mother's education dataframe off of the existing father education dataframe
mother_edu = father_edu_clean.copy()

In [39]:
# Check to visually assess
mother_edu.head()

Country  Student ID  Gender Mother - Highest Schooling  \
4854888   Vietnam        4948  Female           <ISCED level 1>    
4854895   Vietnam        4955  Female           <ISCED level 2>    
4427204    Brazil        2529    Male           <ISCED level 1>    
4490059  Colombia        6551  Female           <ISCED level 1>    
4490057  Colombia        6549  Female          <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
4854888                                    No   
4854895                                    No   
4427204                                    No   
4490059                                    No   
4490057                                   NaN   

        Mother Qualifications - ISCED level 5A  \
4854888                                     No   
4854895                                     No   
4427204                                     No   
4490059                                     No   
4490057                                    Yes   

        Mother Qualifications - ISCED level 5B  \
4854888                                    NaN   
4854895                                    NaN   
4427204                                     No   
4490059                                     No   
4490057                                    NaN   

        Mother Qualifications - ISCED level 4  \
4854888                                    No   
4854895                                    No   
4427204                                    No   
4490059                                   NaN   
4490057                                   NaN   

         Out-of-School Study Time - Homework  \
4854888                                  2.0   
4854895                                  5.0   
4427204                                  2.0   
4490059                                  NaN   
4490057                                  4.0   

         Out-of-School Study Time - Guided Homework  ...  \
4854888                                         2.0  ...   
4854895                                         NaN  ...   
4427204                                         2.0  ...   
4490059                                         NaN  ...   
4490057                                         NaN  ...   

         Out-of-School Study Time - Commercial Company  \
4854888                                            0.0   
4854895                                            0.0   
4427204                                            3.0   
4490059                                            NaN   
4490057                                            3.0   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
4854888                                     0.0                        315.0   
4854895                                     0.0                          NaN   
4427204                                     1.0                        135.0   
4490059                                     NaN                          NaN   
4490057                                     4.0                        360.0   

         Learning Time - Test Language  Learning Time - Science  \
4854888                            NaN                    450.0   
4854895                            NaN                      NaN   
4427204                          180.0                     90.0   
4490059                            NaN                      NaN   
4490057                          360.0                    240.0   

         Average Math Score  Average Reading Score  Average Science Score  \
4854888           441.04222              469.12348              494.90860   
4854895           486.22058              472.61846              536.31110   
4427204           406.06790              433.19248              466.84074   
4490059           367.35474              413.52178              372.28646   
4490057           323.50054              407.16730              360.16410   

         Average Total Score  Education - Father  
4854888     

In [40]:
# Compare original dataframe to father dataframe
print(pisa_clean['Mother - Highest Schooling'].value_counts())
print('-'*47)
print(mother_edu['Mother - Highest Schooling'].value_counts())

<ISCED level 3A>                         236993
<ISCED level 3B, 3C>                      83048
<ISCED level 2>                           82614
<ISCED level 1>                           36556
She did not complete <ISCED level 1>      18768
Name: Mother - Highest Schooling, dtype: int64
-----------------------------------------------
<ISCED level 3A>                         500747
<ISCED level 3B, 3C>                     161851
<ISCED level 2>                          142068
<ISCED level 1>                           61226
She did not complete <ISCED level 1>      30053
Name: Mother - Highest Schooling, dtype: int64


In [41]:
# Split column up with dummies
mother_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'She did not complete <ISCED level 1>']] = pd.get_dummies(mother_edu['Mother - Highest Schooling'])

In [42]:
# Rename the columns for consistency and to remove subcategorization of levels
mother_edu.rename({'She did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Mother Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Mother Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [43]:
# Change datatypes from ints into strings for dummy columns
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].astype(str)
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].astype(str)
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].astype(str)
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].astype(str)

In [44]:
# Change values from '1's and 'Yes's to the corresponding level to prepare for the melt
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
mother_edu['<ISCED level 4>'] = mother_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
mother_edu['<ISCED level 5>'] = mother_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
mother_edu['<ISCED level 6>'] = mother_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [45]:
# Test to see if the columns are all correct before melt
mother_edu.sample(10)

Country  Student ID  Gender  \
3577947         Finland        3843  Female   
4191391          Latvia        1410  Female   
952897           Turkey        2018  Female   
3664536           Italy       25517  Female   
2260895          Mexico        3233  Female   
4074092  United Kingdom        1056    Male   
3821286       Singapore          81  Female   
3399470         Albania        1041    Male   
3119003          Greece        4288    Male   
1341001     New Zealand        1426    Male   

                    Mother - Highest Schooling <ISCED level 6>  \
3577947                      <ISCED level 3A>               No   
4191391                      <ISCED level 3A>               No   
952897                   <ISCED level 3B, 3C>               No   
3664536                      <ISCED level 3A>              NaN   
2260895                       <ISCED level 1>               No   
4074092                  <ISCED level 3B, 3C>               No   
3821286                      <ISCED level 3A>               No   
3399470  She did not complete <ISCED level 1>              NaN   
3119003                      <ISCED level 3A>              NaN   
1341001                      <ISCED level 3A>               No   

        <ISCED level 5>  <ISCED level 5>  <ISCED level 4>  \
3577947              No               No               No   
4191391              No               No  <ISCED level 4>   
952897               No  <ISCED level 5>              NaN   
3664536             NaN              NaN  <ISCED level 4>   
2260895              No               No              NaN   
4074092              No               No               No   
3821286              No               No  <ISCED level 4>   
3399470             NaN              NaN              NaN   
3119003             NaN  <ISCED level 5>              NaN   
1341001              No  <ISCED level 5>               No   

         Out-of-School Study Time - Homework  \
3577947                                  1.0   
4191391                                  4.0   
952897                                   NaN   
3664536                                  7.0   
2260895                                  1.0   
4074092                                  NaN   
3821286                                  8.0   
3399470                                  9.0   
3119003                                  7.0   
1341001                                  7.0   

         Out-of-School Study Time - Guided Homework  ...  Average Math Score  \
3577947                                         0.0  ...           476.63968   
4191391                                         NaN  ...           432.86336   
952897                                          NaN  ...           348.03708   
3664536                                         2.0  ...           444.62532   
2260895                                         1.0  ...           335.65198   
4074092                                         NaN  ...           464.95560   
3821286                                         2.0  ...           582.96462   
3399470                                         9.0  ...           473.36812   
3119003                                         0.0  ...           447.35162   
1341001                                         0.0  ...           621.28836   

         Average Reading Score  Average Science Score  Average Total Score  \
3577947              548.47508              502.18202           509.098927   
4191391              517.65584              462.83104           471.116747   
952897               447.04168              401.47332           398.850693   
3664536              555.70332              416.76614           472.364927   
2260895              319.07830              302.44312           319.057800   
4074092              456.52906              440.07832           453.854327   
3821286              573.49586              587.50466           581.321713   
3399470              438.88630              478.49684           4

In [46]:
# Combine all columns into one variable column
mother_edu = pd.melt(mother_edu, 
                     id_vars=['Country','Student ID', 'Gender',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score',
                             'Education - Father'], var_name='Completed_edu - Mother', value_name='Education - Mother')


In [47]:
# Check to see if the columns from the melt were correctly created
mother_edu.sample(10)

Country  Student ID  Gender  \
8446161           Spain       20683  Female   
1579539           Japan        4079    Male   
1799733          Brazil         282    Male   
3457886         Iceland        2621    Male   
3575131         Uruguay         835    Male   
6722176  United Kingdom        7421  Female   
6946648          Poland        1020    Male   
4594983       Australia         881    Male   
8733354           Spain        4157    Male   
8007510     New Zealand        3433    Male   

         Out-of-School Study Time - Homework  \
8446161                                  1.0   
1579539                                  NaN   
1799733                                  NaN   
3457886                                  3.0   
3575131                                  NaN   
6722176                                  NaN   
6946648                                  7.0   
4594983                                  NaN   
8733354                                  NaN   
8007510                                  1.0   

         Out-of-School Study Time - Guided Homework  \
8446161                                         0.0   
1579539                                         NaN   
1799733                                         NaN   
3457886                                         1.0   
3575131                                         NaN   
6722176                                         NaN   
6946648                                         2.0   
4594983                                         NaN   
8733354                                         NaN   
8007510                                         0.0   

         Out-of-School Study Time - Personal Tutor  \
8446161                                        0.0   
1579539                                        NaN   
1799733                                        NaN   
3457886                                        0.0   
3575131                                        NaN   
6722176                                        NaN   
6946648                                        2.0   
4594983                                        NaN   
8733354                                        NaN   
8007510                                        0.0   

         Out-of-School Study Time - Commercial Company  \
8446161                                            0.0   
1579539                                            NaN   
1799733                                            NaN   
3457886                                            0.0   
3575131                                            NaN   
6722176                                            NaN   
6946648                                            1.0   
4594983                                            NaN   
8733354                                            NaN   
8007510                                            0.0   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
8446161                                     0.0                        180.0   
1579539                                     NaN                          NaN   
1799733                                     NaN                          NaN   
3457886                                     1.0                          NaN   
3575131                                     NaN                          NaN   
6722176                                     NaN                          NaN   
6946648                                     0.0                        225.0   
4594983                                     NaN                          NaN   
8733354                                     NaN                          NaN   
8007510                                     0.0                        280.0   

         Learning Time - Test Language  Learning Time - Science  \
8446161                          180.0                     90.0   
1579539                            NaN                      NaN   
1799733                            NaN         

In [48]:
# Remove all cases from melt where the education of the mother is either '0' or 'No'
mother_edu = mother_edu[(mother_edu['Education - Mother'] != '0')] 
mother_edu = mother_edu[(mother_edu['Education - Mother'] != 'No')]

In [49]:
# Check to see if 0's and No's have been removed
mother_edu.sample(70)

Country  Student ID  Gender  \
2240826             Florida (USA)        1368    Male   
325266                      Spain       19872  Female   
8230840                    Mexico        5420    Male   
3723354                    Mexico       25243  Female   
2708551                     Chile        2605  Female   
2964002                   Estonia        2252  Female   
3952463                    Jordan         974  Female   
380857                  Singapore         128    Male   
6566847                    Mexico       10029    Male   
2775473                    Israel        2363  Female   
7285502                    Mexico       29095    Male   
6928308                 Singapore        1185  Female   
876262                      Korea        1145    Male   
3231335                  Slovenia         610  Female   
6620763                     Italy        7920  Female   
6701884                   Ireland        5011    Male   
18644                  Luxembourg        2008    Male   
2872980        Russian Federation        4284    Male   
3628569  United States of America        2673    Male   
3934817                    Mexico        2183    Male   
3663708                 Australia        7695    Male   
310424                    Denmark        2484    Male   
699089                     Greece        5046  Female   
3497734                    Canada       19731    Male   
2705784               Switzerland        1529    Male   
5703743               Macao-China        3010    Male   
2085092      United Arab Emirates        5614  Female   
4742819                  Portugal        3135  Female   
223496                     Canada        2080  Female   
10062                    Portugal         273    Male   
...                           ...         ...     ...   
4301735                Montenegro        4653    Male   
8013326                     Qatar       10211    Male   
2124844                Costa Rica        4418    Male   
1844304                     Italy       25882  Female   
663318                     Serbia        2045    Male   
139984                    Ireland        4145  Female   
7800305               Switzerland        8403    Male   
273468                     Canada       17212  Female   
2766174                    Serbia        4022  Female   
7086823                     Spain       19229    Male   
4473904                    Mexico       30309    Male   
1806725                   Belgium        4272    Male   
598349                     Brazil       17704    Male   
3571424                   Uruguay        2955  Female   
102452                 Kazakhstan        1647  Female   
4528027                  Colombia        6449  Female   
64105                    Portugal        5371    Male   
109028                Switzerland        5362    Male   
4172148                    Serbia          72    Male   
2458750               Switzerland        7140  Female   
115835                    Albania        4671    Male   
3248065                  Thailand        4211    Male   
797503                     Mexico        1342  Female   
7436135                     Spain       21563    Male   
3680416                    Israel        4313  Female   
550800             United Kingdom        8256    Male   
6207460                Costa Rica        1142    Male   
4772946                    Mexico        9971    Male   
987468                     Brazil       17190    Male   
8165740           Hong Kong-China        2822  Female   

         Out-of-School Study Time - Homework  \
2240826                                  1.0   
325266                                   NaN   
8230840                                  NaN   
3723354                                  2.0   
2708551                                  2.0   
2964002                                 30.0   
3952463                                 20.0   
380857                                   0.0   
6566847                                  5.0   
2775473    

In [50]:
# Remove 'Mother - Highest Schooling' since it is redundant
mother_edu = mother_edu[(mother_edu['Completed_edu - Mother'] != 'Mother - Highest Schooling')] 

In [51]:
# Check to see if correct values remain
mother_edu['Education - Mother'].value_counts()

<ISCED level 3>    662598
<ISCED level 5>    454810
<ISCED level 4>    164266
<ISCED level 2>    142068
<ISCED level 6>     65007
<ISCED level 1>     61226
<ISCED level 0>     30053
Name: Education - Mother, dtype: int64

In [52]:
# Completed_edu is no longer needed since it duplicates the information in 'Education - Mother'
mother_edu.drop(columns=['Completed_edu - Mother'], axis=1, inplace=True)

In [53]:
# Ensure the column has been dropped
mother_edu.sample(10)

Country  Student ID  Gender  \
1184306         Belgium        1039    Male   
8000103        Thailand        3974    Male   
6933528         Romania        4054    Male   
3546368      Montenegro        2051    Male   
5425960          Mexico        7959  Female   
7026090       Australia       11966    Male   
7079595         Belgium        3527  Female   
4583705          Canada       11222    Male   
3687478  United Kingdom         860    Male   
2509810           Spain        7029    Male   

         Out-of-School Study Time - Homework  \
1184306                                  3.0   
8000103                                  4.0   
6933528                                 14.0   
3546368                                  NaN   
5425960                                  NaN   
7026090                                  1.0   
7079595                                  9.0   
4583705                                  NaN   
3687478                                  3.0   
2509810                                 10.0   

         Out-of-School Study Time - Guided Homework  \
1184306                                         0.0   
8000103                                         1.0   
6933528                                         0.0   
3546368                                         NaN   
5425960                                         NaN   
7026090                                         0.0   
7079595                                         0.0   
4583705                                         NaN   
3687478                                         1.0   
2509810                                         0.0   

         Out-of-School Study Time - Personal Tutor  \
1184306                                        0.0   
8000103                                        0.0   
6933528                                        0.0   
3546368                                        NaN   
5425960                                        NaN   
7026090                                        1.0   
7079595                                        0.0   
4583705                                        NaN   
3687478                                        0.0   
2509810                                        0.0   

         Out-of-School Study Time - Commercial Company  \
1184306                                            0.0   
8000103                                            0.0   
6933528                                            0.0   
3546368                                            NaN   
5425960                                            NaN   
7026090                                            0.0   
7079595                                            0.0   
4583705                                            NaN   
3687478                                            0.0   
2509810                                            0.0   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
1184306                                     2.0                        150.0   
8000103                                     0.0                        250.0   
6933528                                     0.0                        150.0   
3546368                                     NaN                          NaN   
5425960                                     NaN                          NaN   
7026090                                     0.0                        265.0   
7079595                                     0.0                        200.0   
4583705                                     NaN                          NaN   
3687478                                     0.0                          NaN   
2509810                                     0.0                        240.0   

         Learning Time - Test Language  Learning Time - Science  \
1184306                          200.0                      NaN   
8000103                          150.0                    150.0   
6933528                          150.0         

In [54]:
# Check if there are any duplicated values remaining
mother_edu['Student ID'].duplicated().sum()

2572439

In [55]:
# mother_edu.replace(["NaN"], np.nan, inplace = True)
# mother_edu.dropna(inplace = True)

In [56]:
# Change datatypes into category, and put into order from lowest to highest
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].astype('category')
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
mother_edu.sort_values(['Education - Mother'], inplace=True)

# Remove duplicates and keep last, which is the highest level of education for the student's father
mother_edu_clean = mother_edu.drop_duplicates(keep='last')

In [57]:
# Ensure there are no duplicates
mother_edu_clean['Student ID'].duplicated().sum()

1968316

In [58]:
# Check the size of the dataframe - 
mother_edu_clean.shape

(2002122, 17)

In [59]:
# Update pisa_clean with final variable from mother and father education
pisa_clean = mother_edu_clean.copy()

In [60]:
mother_edu_clean.head()

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
9339235      Mexico       22883  Female                                  5.0   
9270076    Colombia        4439  Female                                  2.0   
9313010   Singapore         534  Female                                 10.0   
9341153  Luxembourg         721    Male                                  2.0   
9231275      Jordan        1639  Female                                  NaN   

         Out-of-School Study Time - Guided Homework  \
9339235                                         3.0   
9270076                                         NaN   
9313010                                         2.0   
9341153                                         NaN   
9231275                                         NaN   

         Out-of-School Study Time - Personal Tutor  \
9339235                                        1.0   
9270076                                        NaN   
9313010                                        5.0   
9341153                                        0.0   
9231275                                        NaN   

         Out-of-School Study Time - Commercial Company  \
9339235                                            0.0   
9270076                                            NaN   
9313010                                            3.0   
9341153                                            NaN   
9231275                                            NaN   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
9339235                                     2.0                        900.0   
9270076                                     NaN                          NaN   
9313010                                     0.0                        180.0   
9341153                                     NaN                         90.0   
9231275                                     NaN                          NaN   

         Learning Time - Test Language  Learning Time - Science  \
9339235                          480.0                    480.0   
9270076                            NaN                      NaN   
9313010                          225.0                    225.0   
9341153                           90.0                      NaN   
9231275                            NaN                      NaN   

         Average Math Score  Average Reading Score  Average Science Score  \
9339235           312.82910              321.54066              285.84482   
9270076           273.25908              346.16428              294.70346   
9313010           464.17668              479.52896              423.10706   
9341153           379.73982              401.59584              324.72958   
9231275           330.51098              367.92838              353.63672   

         Average Total Score Education - Father Education - Mother  
9339235           306.738193                NaN    <ISCED level 0>  
9270076           304.708940                NaN    <ISCED level 0>  
9313010           455.604233                NaN    <ISCED level 0>  
9341153           368.688413                NaN    <ISCED level 0>  
9231275           350.692027                NaN    <ISCED level 0>

### 4. Missing values in relevant columns 
('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [61]:
# Change string "NaN" to real nan value, and drop those values
pisa_clean.replace(["NaN"], np.nan, inplace = True)
pisa_clean.dropna(inplace = True)

In [62]:
# Check to see if null values correctly dropped
pisa_clean

Country  Student ID  Gender  \
8633803                Jordan        5208  Female   
8435132                Mexico       27182  Female   
8633906  United Arab Emirates        5353  Female   
8435127                Mexico       27237    Male   
8435121                Mexico       27343    Male   
8633551                Jordan        5821    Male   
8633468                Jordan        5952    Male   
8631736                Jordan        6733  Female   
8631832                Jordan        6131  Female   
8632050                Jordan        6309    Male   
8632480                Jordan        6957  Female   
8435134                Mexico       27177    Male   
8632481                Jordan        6960  Female   
8632971                Jordan        4474    Male   
8633227                Jordan        4274    Male   
8635286  United Arab Emirates        7008    Male   
8635397  United Arab Emirates        7712  Female   
8635633  United Arab Emirates        7537    Male   
8635667  United Arab Emirates        6371  Female   
8634616  United Arab Emirates        4297  Female   
8634815  United Arab Emirates        4794    Male   
8435095                Mexico       28077  Female   
8631223                Jordan        1235    Male   
8435141                Mexico       27156  Female   
8620749                 Italy       10666  Female   
8435198                Mexico       23679    Male   
8435191                Mexico       23708  Female   
8435185                Mexico       23751    Male   
8623926                 Italy       17551    Male   
8435182                Mexico       25457  Female   
...                       ...         ...     ...   
1622162                Canada        7089    Male   
1622161           Switzerland        1356  Female   
1622073                 Qatar        4191  Female   
1622070                Canada       10790    Male   
1621994           Switzerland        1900  Female   
1621990                Brazil       10521    Male   
1621985                Brazil        2444    Male   
1621984                 Qatar         937    Male   
1621980                Brazil        9631    Male   
1621979                Brazil        9632    Male   
1621972                 Qatar         931    Male   
1621962                Brazil        6541  Female   
1621950                Canada       14509  Female   
1621942                Canada       14502  Female   
1621940                Canada       14500  Female   
1621959                Canada       11969  Female   
1622064                Brazil        2543    Male   
1622058           Switzerland        1617  Female   
1622048                Canada        7298    Male   
1622039                Canada       14415    Male   
1621999                Brazil        8793    Male   
1622038                Canada       14413  Female   
1622034                Canada       10818    Male   
1622033                Brazil        4139    Male   
1622028                Canada        7343    Male   
1622027                Canada       11942    Male   
1622023                 Qatar         832    Male   
1622014                Brazil        4154    Male   
1622012                Brazil        4153    Male   
1622036                Brazil        8787    Male   

         Out-of-School Study Time - Homework  \
8633803                                  5.0   
8435132                                  4.0   
8633906                                  4.0   
8435127                                  4.0   
8435121                                  4.0   
8633551                                  2.0   
8633468                                  2.0   
8631736                                  4.0   
8631832                                  7.0   
8632050                                  5.0   
8632480                                  2.0   
8435134                                  4.0   
8632481                                  1.0   
8632971                                  1.0   
8633227        

In [63]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

(560077, 17)

### 5. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [64]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning Time - Total'] = pisa_clean['Learning Time - Mathematics'] + pisa_clean['Learning Time - Test Language'] + pisa_clean['Learning Time - Science'] 
                                                                                                                 

In [65]:
list(pisa_clean)

['Country',
 'Student ID',
 'Gender',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'Education - Father',
 'Education - Mother',
 'Out-of-School Study Time - Total',
 'Learning Time - Total']

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [66]:
pisa_clean

Country  Student ID  Gender  \
8633803                Jordan        5208  Female   
8435132                Mexico       27182  Female   
8633906  United Arab Emirates        5353  Female   
8435127                Mexico       27237    Male   
8435121                Mexico       27343    Male   
8633551                Jordan        5821    Male   
8633468                Jordan        5952    Male   
8631736                Jordan        6733  Female   
8631832                Jordan        6131  Female   
8632050                Jordan        6309    Male   
8632480                Jordan        6957  Female   
8435134                Mexico       27177    Male   
8632481                Jordan        6960  Female   
8632971                Jordan        4474    Male   
8633227                Jordan        4274    Male   
8635286  United Arab Emirates        7008    Male   
8635397  United Arab Emirates        7712  Female   
8635633  United Arab Emirates        7537    Male   
8635667  United Arab Emirates        6371  Female   
8634616  United Arab Emirates        4297  Female   
8634815  United Arab Emirates        4794    Male   
8435095                Mexico       28077  Female   
8631223                Jordan        1235    Male   
8435141                Mexico       27156  Female   
8620749                 Italy       10666  Female   
8435198                Mexico       23679    Male   
8435191                Mexico       23708  Female   
8435185                Mexico       23751    Male   
8623926                 Italy       17551    Male   
8435182                Mexico       25457  Female   
...                       ...         ...     ...   
1622162                Canada        7089    Male   
1622161           Switzerland        1356  Female   
1622073                 Qatar        4191  Female   
1622070                Canada       10790    Male   
1621994           Switzerland        1900  Female   
1621990                Brazil       10521    Male   
1621985                Brazil        2444    Male   
1621984                 Qatar         937    Male   
1621980                Brazil        9631    Male   
1621979                Brazil        9632    Male   
1621972                 Qatar         931    Male   
1621962                Brazil        6541  Female   
1621950                Canada       14509  Female   
1621942                Canada       14502  Female   
1621940                Canada       14500  Female   
1621959                Canada       11969  Female   
1622064                Brazil        2543    Male   
1622058           Switzerland        1617  Female   
1622048                Canada        7298    Male   
1622039                Canada       14415    Male   
1621999                Brazil        8793    Male   
1622038                Canada       14413  Female   
1622034                Canada       10818    Male   
1622033                Brazil        4139    Male   
1622028                Canada        7343    Male   
1622027                Canada       11942    Male   
1622023                 Qatar         832    Male   
1622014                Brazil        4154    Male   
1622012                Brazil        4153    Male   
1622036                Brazil        8787    Male   

         Out-of-School Study Time - Homework  \
8633803                                  5.0   
8435132                                  4.0   
8633906                                  4.0   
8435127                                  4.0   
8435121                                  4.0   
8633551                                  2.0   
8633468                                  2.0   
8631736                                  4.0   
8631832                                  7.0   
8632050                                  5.0   
8632480                                  2.0   
8435134                                  4.0   
8632481                                  1.0   
8632971                                  1.0   
8633227        

In [67]:
pisa_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 560077 entries, 8633803 to 1622036
Data columns (total 19 columns):
Country                                          560077 non-null object
Student ID                                       560077 non-null int64
Gender                                           560077 non-null object
Out-of-School Study Time - Homework              560077 non-null float64
Out-of-School Study Time - Guided Homework       560077 non-null float64
Out-of-School Study Time - Personal Tutor        560077 non-null float64
Out-of-School Study Time - Commercial Company    560077 non-null float64
Out-of-School Study Time - With Parent           560077 non-null float64
Learning Time - Mathematics                      560077 non-null float64
Learning Time - Test Language                    560077 non-null float64
Learning Time - Science                          560077 non-null float64
Average Math Score                               560077 non-null float64
Average Re

In [68]:
pisa_clean.shape

(560077, 19)

In [69]:
pisa_clean.drop_duplicates(keep='last', inplace=True)

In [70]:
pisa_clean.duplicated().sum()

0

In [71]:
pisa_clean['Student ID'].drop_duplicates(keep='last', inplace=True)

In [72]:
pisa_clean['Student ID'].duplicated().sum()

0

In [73]:
pisa_clean.shape

(560077, 19)

In [75]:
pisa.shape

(485490, 636)

In [76]:
pisa_clean.head()

Country  Student ID  Gender  \
8633803                Jordan        5208  Female   
8435132                Mexico       27182  Female   
8633906  United Arab Emirates        5353  Female   
8435127                Mexico       27237    Male   
8435121                Mexico       27343    Male   

         Out-of-School Study Time - Homework  \
8633803                                  5.0   
8435132                                  4.0   
8633906                                  4.0   
8435127                                  4.0   
8435121                                  4.0   

         Out-of-School Study Time - Guided Homework  \
8633803                                         1.0   
8435132                                         2.0   
8633906                                         2.0   
8435127                                         1.0   
8435121                                         2.0   

         Out-of-School Study Time - Personal Tutor  \
8633803                                        2.0   
8435132                                        1.0   
8633906                                        3.0   
8435127                                        0.0   
8435121                                        0.0   

         Out-of-School Study Time - Commercial Company  \
8633803                                            0.0   
8435132                                            0.0   
8633906                                            2.0   
8435127                                            0.0   
8435121                                            0.0   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
8633803                                     0.0                        225.0   
8435132                                     1.0                        150.0   
8633906                                     1.0                        315.0   
8435127                                     0.0                        275.0   
8435121                                     2.0                        120.0   

         Learning Time - Test Language  Learning Time - Science  \
8633803                          270.0                    315.0   
8435132                          150.0                    180.0   
8633906                          315.0                    270.0   
8435127                          165.0                    275.0   
8435121                          180.0                    180.0   

         Average Math Score  Average Reading Score  Average Science Score  \
8633803           304.33868              333.37590              355.96794   
8435132           369.69154              382.06712              396.06490   
8633906           367.74420              394.85550              384.96830   
8435127           489.88162              538.56808              426.55726   
8435121           432.94126              357.32846              425.90452   

         Average Total Score Education - Father Education - Mother  \
8633803           331.227507    <ISCED level 3>    <ISCED level 0>   
8435132           382.607853    <ISCED level 0>    <ISCED level 0>   
8633906           382.522667    <ISCED level 3>    <ISCED level 0>   
8435127           485.002320    <ISCED level 0>    <ISCED level 0>   
8435121           405.391413    <ISCED level 0>    <ISCED level 0>   

         Out-of-School Study Time - Total  Learning Time - Total  
8633803                               8.0                  810.0  
8435132                               8.0                  480.0  
8633906                              12.0                  900.0  
8435127                               5.0                  715.0  
8435121                               8.0                  480.0

<a id='sav'></a>
## Store

In [74]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')

<a id='conclusion'></a>
## Conclusion

<a id='references'></a>
## References